### ISSUE TITLE
Deleting memory.bin with each checkpoint can cause write amplification

### ISSUE THREAD

**Issue by botter-nidnul** [urbit/urbit#4150]
On flash storage filesystems that don’t have aggressive defaults regarding TRIM (i.e. specifying the discard mount option in `/etc/fstab`) the frequency with which Urbit writes checkpoints can pretty quickly fill up the device with undiscarded unused blocks, causing write amplification, reduced write speed, and increased wear.

Memory.bin is created, then deleted, every two minutes. The size of memory.bin varies, but it doesn’t have to be particularly large: if each memory.bin averages 100MB then that's 144 GB in 24 hours. The most common behavior by Linux distributions seems to be running fstrim on a weekly basis, so any device less than a terabyte in size risks being filled. And then there’s something like a Raspberry Pi, which typically has much less storage and by default doesn’t run TRIM at all…

Decreasing checkpoint frequency would help a lot, but I think the ideal fix would be to avoid deleting memory.bin and overwriting the file in place instead.

@joemfb

**Comment by ohAitch** [comment]
Presumably it deletes the memory.bin *after* writing the new one, to ensure durability? Which I'm not sure what an in-place alternative to would look like, presumably some manner of generational page table.

**Comment by botter-nidnul** [comment]
The lifespan of a memory.bin file is about 1/2 a second. [Looks to me](https://gist.githubusercontent.com/botter-nidnul/7fa98d305b1a88b41fcd46e946f71877/raw/8b14107583c3ab3e1771f3fc50fa20c67ecc31e7/inotify_output.txt) (warning, 1MB log) like it's created, patch is applied to north.bin, then it's deleted. Then the process is repeated ~119.5 seconds later.

**Comment by joemfb** [comment]
The checkpoint system is conceptually a single memory image of a contiguous arena (the *loom*), implemented by way of four files with conceptually confusing names.

`north.bin` and `south.bin` (no relationship to north/south roads) are the low and high memory segments, respectively. They are created on boot (or restart, if missing), and only every modified by way of patch files. In a logical patch, `memory.bin` contains updated pages, and `control.bin` contains metadata (number of pages, new low/high segment boundaries, checksums). The patch files are created when a (logically) new checkpoint is requested, applied to the memory segments, and then deleted. If patch files exist when we start process, we attempt to apply them, then delete them (regardless of whether we were able to apply them, we'll just replay from the log if the checkpoint is stale).

We need both north/south segments due to the design of a road (reversible stack/heap) on the loom (contiguous arena). The metadata about the road itself (ie, the `u3a_road` struct) lives on road's stack, while all the data is of course on the heap. We only ever persist the home road (which is a north road: heap low, stack high), but we of course need both its data and metadata. And there's usually quite a bit of unused space between them, which we don't want to write.

Long story short: checkpoints are requested every two minutes, which is much too often. And it makes sense that writing and immediately deleting `memory.bin` is suboptimal. I'll have to think through the details for crash handling (ie, do we still write and delete `control.bin`, which is fairly small, and ignore `memory.bin` if control.bin is missing?).

**Comment by ohAitch** [comment]
One off-the-wall idea I have had is splitting loom into 2MB(hugepage) chunks, dumping them into a gitlike two-level hash-named directory structure, and then the only thing you _really_ need to atomically commit is the index table - which at 32kb can go in as a shadow event log entry, probably a dedicated lmdb table acting as a circular buffer of the last e.g. 10. Gets you the north-south split for "free", and you can gc not-recently-checkpointed pages as a nonblocking background operation.
<details><summary>(Might want something more sophisticated than a ring buffer)</summary>
e.g. a logarithmic backoff scheme where there is one snapshot available for each of the last 5 minutes, the previous 5 minutes, preceeding 15, prior 30, hour before that, 2-4 hours ago, etc. I think there's some math for this where every time you add an entry you still delete exactly one. Probably still keep the literal index entries for replay verification, just mark them dead and unlink the pages.</details>

Not sure this would help any on the TRIM front however, and might run you into `vm.max_map_count` issues unless you further consolidate infrequently changed chunks into anonymous regions or sth
